### DISCUSSION TITLE
Pseudoreplicates within biological samples?

### DISCUSSION THREAD

**Discussion by chriswzou** [cancerbits/DElegate#2] [Q&A]
Hi! Thanks for putting this package together. I've found it immensely useful.

I'm running `DElegate::findDE` on a Seurat object that is based off 86 patient samples. To stay truer to the biology, I want to create 3-cell pseudoreplicates within the samples. For example, if I had samples A and B, with 90 cells each, I'd want to have 30 pseudoreplicates per sample, with the cells randomly selected from within the samples only.

As far as I can tell, the `replicate_column` provides a way to assign a replicate identifier to each cell, such that we can achieve what I want like this, where `data` is a Seurat object and `find_deseq_differential_genes` is a loose wrapper around `findDE`.
```
metadata <- data@meta.data
edited_metadata <- metadata %>%
  group_by(sample) %>%
  mutate(
    subgroup_id = ceiling(row_number() / 3),
    pseudoreplicate_id = paste(sample, subgroup_id, sep = "_")
  ) %>%
  ungroup()
data <- AddMetaData(data, edited_metadata$pseudoreplicate_id, col.name = "pseudoreplicate_id")

deseq_df <- utils.deseq$find_deseq_differential_genes(
  data,
  "Recurrent",
  "Primary",
  "recurrence",
  replicate_column = "pseudoreplicate_id"
)
```
Upon running this, however, I tend to see errors like:
```
Error in estimateSizeFactorsForMatrix(counts(object), locfunc = locfunc,  : 
  every gene contains at least one zero, cannot compute log geometric means
```
I know this error is coming from within DESeq, but as it doesn't show up when I run the same code without a replicate column, I can't help but wonder if it has to do with the way the replicate column is expected to be used. Is this correct usage for what I intend to do?

**Comment by ChristophH** [comment]
Sorry for the late reply, I missed this post before.

DESeq was designed for bulk RNA-seq data. The data from just three cells may be too sparse. I suggest to at least combine 10 cells. In your example you are talking about 90 cells in a sample. If you consider those 90 cells to come from one homogeneous cell type/state, then you could use 3 pseudo-replicates with 30 cells each.

Note that the use of pseudo-replicates should be seen as a fallback option to when biological replicates are not available. If you can use patients as biological replicates, you should do so instead.
